# **TikTok Lab Project**
**Course 2 - Get Started with Python**


**The purpose** of this project is to investigate and understand the data provided. This activity will:

1.   Acquainted with data

2.   Compile summary information about the data

3.   Begin the process of EDA and reveal insights contained in the data

4.   Prepare you for more in-depth EDA, hypothesis testing, and statistical analysis


**The goal** is to construct a dataframe in Python, perform a cursory inspection of the provided dataset, and inform TikTok data team members the findings. 


**The ultimate objective** is to use machine learning to classify videos as either claims or opinions. A good first step towards understanding the data might therefore be examining the `claim_status` variable. Begin by determining how many videos there are for each different claim status.
<br/>

*This activity has three parts:*

**Part 1:** Understand the situation
* How can we best prepare to understand and organize the provided TikTok information?

**Part 2:** Understand the data

* Create a pandas dataframe for data learning and future exploratory data analysis (EDA) and statistical activities

* Compile summary information about the data to inform next steps

**Part 3:** Understand the variables

* Use insights from our examination of the summary data to guide deeper investigation into variables

# **Identify data types and compile summary information**


Throughout these project notebooks, we will see references to the problem-solving framework PACE. The following notebook components are labeled with the respective PACE stage: Plan, Analyze, Construct, and Execute.

# **PACE stages**

<img src="images/Pace.png" width="100" height="100" align=left>

   *        [Plan](#scrollTo=psz51YkZVwtN&line=3&uniqifier=1)
   *        [Analyze](#scrollTo=mA7Mz_SnI8km&line=4&uniqifier=1)
   *        [Construct](#scrollTo=Lca9c8XON8lc&line=2&uniqifier=1)
   *        [Execute](#scrollTo=401PgchTPr4E&line=2&uniqifier=1)

<img src="images/Plan.png" width="100" height="100" align=left>


## **PACE: Plan**

Consider the questions in your PACE Strategy Document and those below to craft your response:



### **Task 1. Understand the situation**

We will start by reading in the data, viewing the data dictionary and exploring the data set to identify key variables for the stakeholders.


<img src="images/Analyze.png" width="100" height="100" align=left>

## **PACE: Analyze**

Consider the questions in your PACE Strategy Document to reflect on the Analyze stage.

### **Task 2a. Imports and data loading**

Start by importing the packages to load and explore the dataset. Make sure to use the following import statements:

In [3]:
# Import packages
import pandas as pd
import numpy as np

Then, load the dataset into a dataframe. Creating a dataframe will help conduct data manipulation, exploratory data analysis (EDA), and statistical activities.

In [4]:
# Load dataset into dataframe
data = pd.read_csv("tiktok_dataset.csv")

### **Task 2b. Understand the data - Inspect the data**

We will view and inspect summary information about the dataframe by **coding the following:**

In [5]:
# Display and examine the first ten rows of the dataframe
data.head(3)

# claim_status    video_id  video_duration_sec  \
0  1        claim  7017666017                  59   
1  2        claim  4014381136                  32   
2  3        claim  9859838091                  31   

                            video_transcription_text verified_status  \
0  someone shared with me that drone deliveries a...    not verified   
1  someone shared with me that there are more mic...    not verified   
2  someone shared with me that american industria...    not verified   

  author_ban_status  video_view_count  video_like_count  video_share_count  \
0      under review          343296.0           19425.0              241.0   
1            active          140877.0           77355.0            19034.0   
2            active          902185.0           97690.0             2858.0   

   video_download_count  video_comment_count  
0                   1.0                  0.0  
1                1161.0                684.0  
2                 833.0                329.0

In [6]:
# Get summary info
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19382 entries, 0 to 19381
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   #                         19382 non-null  int64  
 1   claim_status              19084 non-null  object 
 2   video_id                  19382 non-null  int64  
 3   video_duration_sec        19382 non-null  int64  
 4   video_transcription_text  19084 non-null  object 
 5   verified_status           19382 non-null  object 
 6   author_ban_status         19382 non-null  object 
 7   video_view_count          19084 non-null  float64
 8   video_like_count          19084 non-null  float64
 9   video_share_count         19084 non-null  float64
 10  video_download_count      19084 non-null  float64
 11  video_comment_count       19084 non-null  float64
dtypes: float64(5), int64(3), object(4)
memory usage: 1.8+ MB


In [7]:
# Get summary statistics
data.describe()

#      video_id  video_duration_sec  video_view_count  \
count  19382.000000  1.938200e+04        19382.000000      19084.000000   
mean    9691.500000  5.627454e+09           32.421732     254708.558688   
std     5595.245794  2.536440e+09           16.229967     322893.280814   
min        1.000000  1.234959e+09            5.000000         20.000000   
25%     4846.250000  3.430417e+09           18.000000       4942.500000   
50%     9691.500000  5.618664e+09           32.000000       9954.500000   
75%    14536.750000  7.843960e+09           47.000000     504327.000000   
max    19382.000000  9.999873e+09           60.000000     999817.000000   

       video_like_count  video_share_count  video_download_count  \
count      19084.000000       19084.000000          19084.000000   
mean       84304.636030       16735.248323           1049.429627   
std       133420.546814       32036.174350           2004.299894   
min            0.000000           0.000000              0.000000   
25%          810.750000         115.000000              7.000000   
50%         3403.500000         717.000000             46.000000   
75%       125020.000000       18222.000000           1156.250000   
max       657830.000000      256130.000000          14994.000000   

       video_comment_count  
count         19084.000000  
mean            349.312146  
std             799.638865  
min               0.000000  
25%               1.000000  
50%               9.000000  
75%             292.000000  
max            9599.000000

# Obeservations:

- The dataframe contains a collection of categorical, text, and numerical data. Each row represents a distinct TikTok video presenting a claim or an opinion, along with accompanying metadata.

- The data frame comprises 5 floats, 3 integers, and 4 objects. There are 19,382 observations, with 298 missing values in claim_status, video_transcription, video_like_count, video_share_count, video_download_count, and video_comment_count.

- Many of the count variables appear to have outliers at the high end of their distributions. They exhibit large standard deviations and maximum values that significantly exceed their quartile values.

### **Task 2c. Understand the data - Investigate the variables**

In this phase, we will begin to investigate the variables more closely to better understand them.

Ultimate objective is to use machine learning to classify videos as either claims or opinions. A good first step towards understanding the data might therefore be examining the `claim_status` variable. Begin by determining how many videos there are for each different claim status.

In [8]:
# What are the different values for claim status and how many of each are in the data?
data['claim_status'].value_counts()


claim      9608
opinion    9476
Name: claim_status, dtype: int64

**Observation** - balanced distribution between cliam and opinion, with 298 having no values

Next, examine the engagement trends associated with each different claim status.

Start by using Boolean masking to filter the data according to claim status, then calculate the mean and median view counts for each claim status.

In [9]:
claims= data[data['claim_status'] == 'claim']

print ("Mean View count claims: {:.2f}".format(claims['video_view_count'].mean()))
print ("Median View count claims: {:.2f}".format(claims['video_view_count'].median()))


Mean View count claims: 501029.45
Median View count claims: 501555.00


In [10]:
# What is the average view count of videos with "opinion" status?
# What is the average view count of videos with "claim" status?
claims= data[data['claim_status'] == 'opinion']

print ("Mean View count claims: {:.2f}".format(claims['video_view_count'].mean()))
print ("Median View count claims: {:.2f}".format(claims['video_view_count'].median()))

Mean View count claims: 4956.43
Median View count claims: 4953.00


The mean and the median within each claim category are close to one another, but there is a vast discrepancy between view counts for videos labeled as claims and videos labeled as opinions. 50 k vs 5k

Now, examine trends associated with the ban status of the author.

Use `groupby()` to calculate how many videos there are for each combination of categories of claim status and author ban status.

In [11]:
# Get counts for each group combination of claim status and author ban status
### YOUR CODE HERE ###
# data.groupby(['claim_status','author_ban_status']).count()[['#','video_id']]
(data.groupby(['claim_status','author_ban_status'])
    .agg({'#':'count'})
    .rename(columns={'#':'Count of Index'})
                                                         
) 

# print (data['author_ban_status'].value_counts())
# print (data['author_ban_status'].value_counts().sum())

Count of Index
claim_status author_ban_status                
claim        active                       6566
             banned                       1439
             under review                 1603
opinion      active                       8817
             banned                        196
             under review                  463

**Question:** What do you notice about the number of claims videos with banned authors? Why might this relationship occur?

**Obesrvation:** 15% of total of claim videos are banned vs 2% of total of opinion videos are banner.  

- Claim videos are more strictly policed the opinion videos
- Authors must comply with a stricter set of riles if they post a claim than if they post a opinion

Also it should be noted that there is no way of knowing if clain video are inherently more likely than opinion videos to result in authoer bans. of if authors who post claim video are more likely to post video that violates terms of service. 

Lets  Continue investigating engagement levels, now focusing on `author_ban_status`. what is the median video share count of each author ban status.

In [12]:
### YOUR CODE HERE ###
data.groupby(['author_ban_status']).agg({
                                        'video_view_count': ['mean','median'],
                                        'video_like_count': ['mean','median'],
                                        'video_share_count': ['mean','median'],
    
    
     
                                        }) 

video_view_count           video_like_count            \
                              mean    median             mean    median   
author_ban_status                                                         
active               215927.039524    8616.0     71036.533836    2222.0   
banned               445845.439144  448201.0    153017.236697  105573.0   
under review         392204.836399  365245.5    128718.050339   71204.5   

                  video_share_count           
                               mean   median  
author_ban_status                             
active                 14111.466164    437.0  
banned                 29998.942508  14468.0  
under review           25774.696999   9444.0

In [13]:
# What's the median video share count of each author ban status?
### YOUR CODE HERE ###
data.groupby(['author_ban_status']).median(numeric_only=True)[['video_share_count']]

video_share_count
author_ban_status                   
active                         437.0
banned                       14468.0
under review                  9444.0

**Question:** What do you notice about the share count of banned authors, compared to that of active authors? Explore this in more depth.

**Obeservation** view like and share count medians for banned authors are 52 47 33 times higher vs active authors 


Use `groupby()` to group the data by `author_ban_status`, then use `agg()` to get the count, mean, and median of each of the following columns:
* `video_view_count`
* `video_like_count`
* `video_share_count`

Remember, the argument for the `agg()` function is a dictionary whose keys are columns. The values for each column are a list of the calculations you want to perform.

In [14]:
### YOUR CODE HERE ###


data.groupby(['author_ban_status']).agg(
    {'video_view_count': ['count', 'mean', 'median'],
     'video_like_count': ['count', 'mean', 'median'],
     'video_share_count': ['count', 'mean', 'median']
     })

video_view_count                          video_like_count  \
                             count           mean    median            count   
author_ban_status                                                              
active                       15383  215927.039524    8616.0            15383   
banned                        1635  445845.439144  448201.0             1635   
under review                  2066  392204.836399  365245.5             2066   

                                           video_share_count                \
                            mean    median             count          mean   
author_ban_status                                                            
active              71036.533836    2222.0             15383  14111.466164   
banned             153017.236697  105573.0              1635  29998.942508   
under review       128718.050339   71204.5              2066  25774.696999   

                            
                    median  
author_ban_status           
active               437.0  
banned             14468.0  
under review        9444.0

**Question:** What do you notice about the number of views, likes, and shares for banned authors compared to active authors?

Now, create three new columns to help better understand engagement rates:
* `likes_per_view`: represents the number of likes divided by the number of views for each video
* `comments_per_view`: represents the number of comments divided by the number of views for each video
* `shares_per_view`: represents the number of shares divided by the number of views for each video

In [23]:
# Create a likes_per_view column
data['likes_per_view'] = data['video_like_count']/ data ['video_view_count']

# Create a comments_per_view column
data['comments_per_view'] = data['video_comment_count']/ data ['video_view_count']

# Create a shares_per_view column
data['shares_per_view'] = data['video_share_count']/ data ['video_view_count']

Use `groupby()` to compile the information in each of the three newly created columns for each combination of categories of claim status and author ban status, then use `agg()` to calculate the count, the mean, and the median of each group.

In [30]:
### YOUR CODE HERE ###
data.groupby(['claim_status','author_ban_status']).agg({
    'likes_per_view': ['mean','median'],
    'comments_per_view': ['mean','median'],
    'shares_per_view': ['mean','median']
})

likes_per_view           comments_per_view  \
                                         mean    median              mean   
claim_status author_ban_status                                              
claim        active                  0.329542  0.326538          0.001393   
             banned                  0.345071  0.358909          0.001377   
             under review            0.327997  0.320867          0.001367   
opinion      active                  0.219744  0.218330          0.000517   
             banned                  0.206868  0.198483          0.000434   
             under review            0.226394  0.228051          0.000536   

                                         shares_per_view            
                                  median            mean    median  
claim_status author_ban_status                                      
claim        active             0.000776        0.065456  0.049279  
             banned             0.000746        0.067893  0.051606  
             under review       0.000789        0.065733  0.049967  
opinion      active             0.000252        0.043729  0.032405  
             banned             0.000193        0.040531  0.030728  
             under review       0.000293        0.044472  0.035027


**Obeservation** Videos by banned authors and those under review tend to get far more views, likes, and shares than videos by non-banned authors. However, *when a video does get viewed*, its engagement rate is less related to author ban status and more related to its claim status.

Also, we know that claim videos have a higher view rate than opinion videos, but this tells us that claim videos also have a higher rate of likes on average, so they are more favorably received as well. Furthermore, they receive more engagement via comments and shares than opinion videos.

Note that for claim videos, banned authors have slightly higher likes/view and shares/view rates than active authors or those under review. However, for opinion videos, active authors and those under review both get higher engagement rates than banned authors in all categories.


<img src="images/Construct.png" width="100" height="100" align=left>

## **PACE: Construct**

**Note**: The Construct stage does not apply to this workflow. The PACE framework can be adapted to fit the specific requirements of any project.




<img src="images/Execute.png" width="100" height="100" align=left>

## **PACE: Execute**

Consider the questions in your PACE Strategy Document and those below to craft your response.

### **Given your efforts, what can you summarize for Rosie Mae Bradshaw and the TikTok data team?**

Of the 19,382 samples in this dataset, just over 50% are claims—9,608 of them.
Engagement level is strongly correlated with claim status. This should be a focus of further inquiry.
Videos with banned authors have significantly higher engagement than videos with active authors. Videos with authors under review fall between these two categories in terms of engagement levels.